### PR TITLE
docs: API specification for CRDs

### LINKED ISSUE #69: API specification for CRDs
We need API references for our CRDs, like Istio, Service APIs.

* https://istio.io/latest/docs/reference/config/networking/virtual-service/
* https://kubernetes-sigs.github.io/service-apis/spec/

### PR DESCRIPTION
**Pull request by Firstsawyou**
close #69

### CODE REVIEW

File: docs/API-specification.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
+<!--
```
Comment (tokers) [inline docs/API-specification.md:1]: I think CRD-specification is more appropriate.

File: docs/API-specification.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+# API specification
+
+In order to be able to use yaml to define the objects required by apisix in k8s, the following structure is defined.
```
Comment (tokers) [inline docs/API-specification.md:22]: What about "In order to control the behavior of the proxy (Apache APISIX), the following CRDs should be defined".

Also, always "Apache APISIX".

File: docs/API-specification.md [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+## ApisixRoute
+
+`ApisixRoute` corresponds to the `route` object in apisix.
```
Comment (tokers) [inline docs/API-specification.md:26]: That's not really right, you may try to compare the `Route` in APISIX and `ApisixRoute`.

File: docs/API-specification.md [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+          servicePort: 8080
+        path: /hello*
+        plugins:
+          - httpserver-plugins
+          - ...      
+```
+
+|     Field     |  Type    | Description    |
+|---------------|----------|--------------------------------|
+| apiVersion    | string   | apisix.apache.org/v1           |
```
Comment (tokers) [inline docs/API-specification.md:55]: Well, you don't need to explain fields such as `apiVersion`, it's not defined by us, it's common field in Kubernetes Resources.

File: docs/API-specification.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  port: 8080
+  plugins:
+    - httpserver-plugins
+    - ...
+```
+
+|     Field     |  Type    | Description    |
+|---------------|----------|----------------|
+| apiVersion    | string   | apisix.apache.org/v1    |
+| kind          | string   | ApisixService           |
```
Comment (tokers) [inline docs/API-specification.md:83]: Ditto.

File: docs/API-specification.md [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+  	active:
+  		...
+  	passive:
+  		...
+```
+
+|     Field     |  Type    | Description    |
+|---------------|----------|----------------|
+| apiVersion    | string   | apisix.apache.org/v1    |
+| kind          | string   | ApisixUpstream          |
```
Comment (tokers) [inline docs/API-specification.md:109]: Ditto.

File: docs/API-specification.md [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+          - httpserver-plugins
+          - ...      
+```
+
+|     Field     |  Type    | Description    |
+|---------------|----------|--------------------------------|
+| apiVersion    | string   | apisix.apache.org/v1           |
+| kind          | string   | ApisixRoute         |
+| name          | string   | httpserverRoute         |
+| namespace     | string   | Specify `namespace`, only one `backend` under `namespace` can be configured in the same yaml file. |
```
Comment (tokers) [inline docs/API-specification.md:58]: You should, however, explain each fields in `spec`.

**Review by tokers — CHANGES_REQUESTED** [review]

File: docs/API-specification.md (hunk 1 above)
Comment (Firstsawyou) [reply docs/API-specification.md:1]: +1，agree

File: docs/API-specification.md (hunk 4 above)
Comment (Firstsawyou) [reply docs/API-specification.md:55]: @tokers 
Where can I view all the fields defined by us, currently I refer to the fields defined here: [https://github.com/apache/apisix-ingress-controller/issues/3 ](https://github.com/apache/apisix-ingress-controller/issues/3) .

File: docs/API-specification.md (hunk 7 above)
Comment (Firstsawyou) [reply docs/API-specification.md]: updated.

File: docs/API-specification.md (hunk 2 above)
Comment (Firstsawyou) [reply docs/API-specification.md]: updated.

File: docs/CRD-specification.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+# CRD specification
+
+In order to control the behavior of the proxy (Apache APISIX), the following CRDs should be defined
```
Comment (tokers) [inline docs/CRD-specification.md:22]: Style: miss `.` at the end of line.

File: docs/CRD-specification.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+```
+
+|     Field     |  Type    | Description                     |
+|---------------|----------|--------------------------------|
+| rules         | array    | ApisixRoute's request matching rules. |
+| host          | string   | The requested host.        |
+| http          | string   | The routing rule of the request.         |
+| paths         | array    | The routing rule of the request.      |
+| backend       | object   | Backend service information configuration.         |
+| serviceName   | string   | The name of the service.        |
```
Comment (tokers) [inline docs/CRD-specification.md:73]: The name of backend service.

File: docs/CRD-specification.md [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+| servicePort   | int      | The port of the service.        |
```
Comment (tokers) [inline docs/CRD-specification.md:74]: The port of backend service.

File: docs/CRD-specification.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+| path          | string   | The URI matched by the route. Supports exact match and prefix match. Example，exact match: `/hello`, prefix match: `/hello/*`.   |
+| plugins       | array    | Custom plugin collection (Plugins defined in the `route` level).    |
```
Comment (tokers) [inline docs/CRD-specification.md:76]: Add a hyperlink to apisix plugins?

File: docs/CRD-specification.md [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+| name          | string   | The name of the plugin.      |
+| enable        | boolean  | Whether to enable the plugin, `true`: means enable, `false`: means disable.      |
+| config        | object   |  Configuration of plugin information.    |
```
Comment (tokers) [inline docs/CRD-specification.md:79]: Add a note to say the check of configuration schema is missing now, so please be careful when editing.

File: docs/CRD-specification.md [hunk 13]
```diff
@@ -0,0 +1,198 @@
...
+
+- [ApisixRoute](#apisixroute)
+- [ApisixService](#apisixservice)
+- [ApisixUpstream](#apisixupstream)
+- [ApisixPlugin](#apisixplugin)
+- [ApisixTls](#apisixtls)
+
+## ApisixRoute
+
+`ApisixRoute` corresponds to the `route` object in Apache APISIX.
```
Comment (tokers) [inline docs/CRD-specification.md:34]: It's too concise, you should add some descriptions for `route`.

Comment (tokers) [reply docs/CRD-specification.md:34]: Also, hyperlink to APISIX route is necessary.

File: docs/CRD-specification.md [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+| servicePort   | int      | The port of the service.        |
+| path          | string   | The URI matched by the route. Supports exact match and prefix match. Example，exact match: `/hello`, prefix match: `/hello/*`.   |
+| plugins       | array    | Custom plugin collection (Plugins defined in the `route` level).    |
+| name          | string   | The name of the plugin.      |
+| enable        | boolean  | Whether to enable the plugin, `true`: means enable, `false`: means disable.      |
+| config        | object   |  Configuration of plugin information.    |
+
+## ApisixService
+
+`ApisixService` corresponds to the `service` object in Apache APISIX.
```
Comment (tokers) [inline docs/CRD-specification.md:83]: Add a link to APISIX Service.

File: docs/CRD-specification.md [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+        time_window: 60
+        rejected_code: 503
+        key: remote_addr
+```
+
+|     Field     |  Type    | Description    |
+|---------------|----------|----------------|
+| upstream      | string   | The name of the upstream service.    |
+| port          | int      | The port number of the upstream service.    |
+| plugins       | array   | Custom plugin collection (Plugins defined in the `service` level).  |
```
Comment (tokers) [inline docs/CRD-specification.md:110]: Add an anchor to the fields description in `ApisixRoute`.

File: docs/CRD-specification.md [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+
+## ApisixUpstream
+
+`ApisixUpstream` corresponds to the `upstream` object in Apache APISIX.
```
Comment (tokers) [inline docs/CRD-specification.md:114]: Ditto, add a link to APISIX Upstream.

File: docs/CRD-specification.md [hunk 17]
```diff
@@ -0,0 +1,198 @@
...
+  ports:
+    - port: 8080
+      loadbalancer: roundrobin
+```
+
+|     Field     |  Type    | Description    |
+|---------------|----------|----------------|
+| ports         | array    | Custom upstream collection.   |
+| port          | int      | Upstream service port number.    |
+| loadbalancer  | string/object   | Select the algorithm of the upstream service, support `roundrobin` and `chash`.  |
```
Comment (tokers) [inline docs/CRD-specification.md:134]: Bad description. What about:

```
The load balance algorithm of this upstream service, optional value can be "roundrobin" and "chash".
```

File: docs/CRD-specification.md [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+
+## ApisixPlugin
```
Comment (tokers) [inline docs/CRD-specification.md:136]: We don't have this CRD.

File: docs/CRD-specification.md [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+```
+
+|     Field     |  Type    | Description                 |
+|---------------|----------|-----------------------------|
+| plugins       | array   | Custom plugin collection.    |
+| name          | string   | The name of the plugin.      |
+| enable        | boolean  | Whether to enable the plugin, `true`: means enable, `false`: means disable.  |
+| config        | object   |  Configuration of plugin information. |
+
+## ApisixTls
```
Comment (tokers) [inline docs/CRD-specification.md:175]: No description?

File: docs/CRD-specification.md [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+  - asr.duiopen.com
+  - tts.duiopen.com
+  secret:
+    name: all.duiopen.com
+    namespace: cloud
+```
+
+|     Field     |  Type    | Description                     |
+|---------------|----------|---------------------------------|
+| hosts         | array    | Host of `Tls`.                  |
```
Comment (tokers) [inline docs/CRD-specification.md:195]: It should be `SNI` in TLS.

**Review by tokers — CHANGES_REQUESTED** [review]

File: docs/CRD-specification.md (hunk 18 above)
Comment (Firstsawyou) [reply docs/CRD-specification.md:136]: I saw that this CRD design was included, so I added it.
[https://github.com/apache/apisix-ingress-controller/issues/3 .](https://github.com/apache/apisix-ingress-controller/issues/3).

Do we need to delete this CRD?

File: docs/CRD-specification.md (hunk 19 above)
Comment (Firstsawyou) [reply docs/CRD-specification.md:175]: Forgot to add, I will update later.

File: docs/CRD-specification.md (hunk 12 above)
Comment (Firstsawyou) [reply docs/CRD-specification.md]: updated.

File: docs/CRD-specification.md [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+# CRD specification
+
+In order to control the behavior of the proxy ([Apache APISIX](https://github.com/apache/apisix)), the following CRDs should be defined.
```
Comment (gxthrj) [inline docs/CRD-specification.md:22]: > the following CRDs should be defined.

It seems that each has to be defined. 
Suggest to change to `Use the following CRDs to define your routing rules. Normally, you only need to use ApisixRoute to complete a simple route definition.`

File: docs/CRD-specification.md [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+|     Field     |  Type    |                    Description                     |
+|---------------|----------|----------------------------------------------------|
+| rules         | array    | ApisixRoute's request matching rules.              |
+| host          | string   | The requested host.                                |
+| http          | string   | The routing rule of the request.                   |
+| paths         | array    | The routing rule of the request.                   |
+| backend       | object   | Backend service information configuration.         |
+| serviceName   | string   | The name of backend service.                       |
+| servicePort   | int      | The port of backend service.                       |
+| path          | string   | The URI matched by the route. Supports exact match and prefix match. Example，exact match: `/hello`, prefix match: `/hello/*`.   |
```
Comment (gxthrj) [inline docs/CRD-specification.md:76]: `/hello/*` -> `/hello*`

File: docs/CRD-specification.md [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+To learn more, please check the [Apache APISIX architecture-design document](https://github.com/apache/apisix/blob/master/doc/architecture-design.md#route).
+
+Structure example:
+
+```yaml
+apiVersion: apisix.apache.org/v1
+kind: ApisixRoute
+metadata:
+  annotations:
+    k8s.apisix.apache.org/ingress.class: apisix_group
```
Comment (gxthrj) [inline docs/CRD-specification.md:44]: I suggest to remove `anotations` and describe `annotations` alone.

File: docs/CRD-specification.md [hunk 24]
```diff
@@ -0,0 +1,174 @@
...
+        path: /hello*
+        plugins:
+          - name: limit-conn
+            enable: true
+            config:
+              count: 2
+              time_window: 60
+              rejected_code: 503
+              key: remote_addr
+```
```
Comment (gxthrj) [inline docs/CRD-specification.md:65]: `serviceName` and `servicePort` need to add a description.
`namespace + serviceName + servicePort ` form a unique identifier to match the back-end service.

File: docs/CRD-specification.md [hunk 25]
```diff
@@ -0,0 +1,174 @@
...
+apiVersion: apisix.apache.org/v1
+kind: ApisixService
+metadata:
+  name: httpserver
+  namespace: cloud  
+spec:
+  upstream: httpserver
+  port: 8080
+  plugins:
+    - name: limit-conn
```
Comment (gxthrj) [inline docs/CRD-specification.md:100]: the plugins `limit-conn` need a link to describe the schema. 
This is the `config` filling rule below.

File: docs/CRD-specification.md [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+    name: all.duiopen.com
+    namespace: cloud
+```
+
+|     Field     |  Type    | Description                     |
+|---------------|----------|---------------------------------|
+| hosts         | array    | Host of `SNI`.                  |
+| secret        | object   | Secret information.             |
+| name          | string   | The name of `SNI` secret.         |
+| namespace     | string   | The namespace of `SNI` secret.    |
```
Comment (gxthrj) [inline docs/CRD-specification.md:172]: `secret.namespace` and `secret.name` are the unique identifier to match k8s `secret`.
The  content in `secret` should contains `key` and `cert`.

File: docs/CRD-specification.md (hunk 21 above)
Comment (Firstsawyou) [reply docs/CRD-specification.md]: updated.

File: docs/CRD-specification.md (hunk 22 above)
Comment (Firstsawyou) [reply docs/CRD-specification.md]: updated.

File: docs/CRD-specification.md (hunk 23 above)
Comment (Firstsawyou) [reply docs/CRD-specification.md]: done.

File: docs/CRD-specification.md [hunk 27]
```diff
@@ -0,0 +1,196 @@
...
+
+|     Field     |  Type    | Description    |
+|---------------|----------|----------------|
+| ports         | array    | Custom upstream collection.   |
+| port          | int      | Upstream service port number.    |
+| loadbalancer  | string/object   | The load balance algorithm of this upstream service, optional value can be `roundrobin` or `chash`.  |
+
+## ApisixTls
+
+`ApisixTls` corresponds to the `SSL` of the `Router` object in Apache APISIX.
```
Comment (tokers) [inline docs/CRD-specification.md:169]: We don't have `Router` object in Apache APISIX.

File: docs/CRD-specification.md [hunk 28]
```diff
@@ -0,0 +1,196 @@
...
+  - tts.duiopen.com
+  secret:
+    name: all.duiopen.com
+    namespace: cloud
+```
+
+|     Field     |  Type    | Description                     |
+|---------------|----------|---------------------------------|
+| hosts         | array    | Host of `SNI`.                  |
+| secret        | object   | The secret of kubernetes, it works with `ApisixTls`.            |
```
Comment (tokers) [inline docs/CRD-specification.md:192]: bad desc.

What about "The definition of the related Secret object with current ApisixTls object".

File: docs/CRD-specification.md [hunk 29]
```diff
@@ -0,0 +1,196 @@
...
+  - asr.duiopen.com
+  - tts.duiopen.com
+  secret:
+    name: all.duiopen.com
+    namespace: cloud
+```
+
+|     Field     |  Type    | Description                     |
+|---------------|----------|---------------------------------|
+| hosts         | array    | Host of `SNI`.                  |
```
Comment (tokers) [inline docs/CRD-specification.md:191]: Unintelligible desc.

What about "The domain list to identify which hosts (matched with SNI) can use the TLS certificate stored in the Secret"

File: docs/CRD-specification.md [hunk 30]
```diff
@@ -0,0 +1,196 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+
+# CRD specification
+
+Use the following CRDs to define your routing rules. Normally, you only need to use ApisixRoute to complete a simple route definition.
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/CRD-specification.md:22]: @gxthrj  @tokers 
How about this description?
"In order to control the behavior of the proxy ([Apache APISIX](https://github.com/apache/apisix)), the following CRDs should be defined. Normally, you only need to use ApisixRoute to complete a simple route definition."

File: docs/CRD-specification.md [hunk 31]
```diff
@@ -0,0 +1,196 @@
...
+              time_window: 60
+              rejected_code: 503
+              key: remote_addr
+```
+
+|     Field     |  Type    |                    Description                     |
+|---------------|----------|----------------------------------------------------|
+| rules         | array    | ApisixRoute's request matching rules.              |
+| host          | string   | The requested host.                                |
+| http          | string   | The routing rule of the request.                   |
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/CRD-specification.md:68]: I wrote this comment according to here([https://github.com/apache/apisix-ingress-controller/issues/3 ](https://github.com/apache/apisix-ingress-controller/issues/3)). How can I describe it more appropriately? @gxthrj @tokers

File: docs/CRD-specification.md [hunk 32]
```diff
@@ -0,0 +1,196 @@
...
+      - 1.2.3.4/16
+      - 4.3.2.1/8
+  name: httpserverRoute
+  namespace: cloud
+spec:
+```
+
+|         Field                                  |    Type    |                       Description                                  |
+|------------------------------------------------|------------|--------------------------------------------------------------------|
+| `k8s.apisix.apache.org/ssl-redirect`           | boolean    | Whether to use SSL forwarding, true: means using SSL forwarding, false: means not using SSL forwarding.   |
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/CRD-specification.md:100]: I wrote this comment according to here([https://github.com/apache/apisix-ingress-controller/issues/3 ](https://github.com/apache/apisix-ingress-controller/issues/3)). How can I describe it more appropriately? @gxthrj @tokers

[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/CRD-specification.md:100]: Whether to force http to be redirected from https. ture: means forced conversion to https, false: means no conversion. How about this description?

[the remark this answers is not in the document]
Comment (tokers) [reply docs/CRD-specification.md:100]: OK, that's more appropriate.

File: docs/CRD-specification.md (hunk 31 above)
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/CRD-specification.md]: done.

File: docs/CRD-specification.md [hunk 33]
```diff
@@ -0,0 +1,196 @@
...
+              rejected_code: 503
+              key: remote_addr
+```
+
+|     Field     |  Type    |                    Description                     |
+|---------------|----------|----------------------------------------------------|
+| rules         | array    | ApisixRoute's request matching rules.              |
+| host          | string   | The requested host.                                |
+| http          | string   | The routing rule of the request.                   |
+| paths         | array    | The routing rule of the request.                   |
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/CRD-specification.md]: updated.

File: docs/CRD-specification.md (hunk 32 above)
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/CRD-specification.md]: done.

File: docs/CRD-specification.md (hunk 27 above)
Comment (Firstsawyou) [reply docs/CRD-specification.md]: fixed

File: docs/CRD-specification.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (gxthrj) [reply docs/CRD-specification.md]: OK , now we have ApisixTls, which is not belong to route. you can remove `Normally, you only need to use ApisixRoute to complete a simple route definition.`

File: docs/CRD-specification.md (hunk 26 above)
Comment (gxthrj) [reply docs/CRD-specification.md]: The desc of `name` and `namespace` can be set like this:
`The name of secret, the secret contains key and cert for TLS`
`The namespace of secret , the secret contains key and cert  for TLS`

**Review by gxthrj — CHANGES_REQUESTED** [review]

File: docs/CRD-specification.md [hunk 34]
```diff
@@ -0,0 +1,196 @@
...
+      - 1.2.3.4/16
+      - 4.3.2.1/8
+  name: httpserverRoute
+  namespace: cloud
+spec:
+```
+
+|         Field                                  |    Type    |                       Description                                  |
+|------------------------------------------------|------------|--------------------------------------------------------------------|
+| `k8s.apisix.apache.org/ssl-redirect`           | boolean    | Whether to force http to be redirected from https. ture: means forced conversion to https, false: means no conversion.   |
```
Comment (tokers) [inline docs/CRD-specification.md:100]: redirect to, not from.

File: docs/CRD-specification.md [hunk 35]
```diff
@@ -0,0 +1,196 @@
...
+|     Field     |  Type    | Description    |
+|---------------|----------|----------------|
+| ports         | array    | Custom upstream collection.   |
+| port          | int      | Upstream service port number.    |
+| loadbalancer  | string/object   | The load balance algorithm of this upstream service, optional value can be `roundrobin` or `chash`.  |
+
+## ApisixTls
+
+`ApisixTls` corresponds to the SSL load matching route in Apache APISIX.
+`SSL` loads the matching route. (Default) Use SNI (Server Name Indication) as the primary index.
```
Comment (tokers) [inline docs/CRD-specification.md:170]: I think this sentence is useless?

Comment (Firstsawyou) [reply docs/CRD-specification.md:170]: Yes, this sentence is in the APISIX doc, it seems we don't need this sentence.

Comment (Firstsawyou) [reply docs/CRD-specification.md]: updated.

File: docs/CRD-specification.md (hunk 34 above)
Comment (Firstsawyou) [reply docs/CRD-specification.md]: done.

File: docs/CRD-specification.md (hunk 30 above)
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply docs/CRD-specification.md]: done.

File: docs/CRD-specification.md (hunk 26 above)
Comment (Firstsawyou) [reply docs/CRD-specification.md]: done.

**Review by tokers — APPROVED** [review]

**Review by gxthrj — APPROVED** [review]

**Pull request merged**